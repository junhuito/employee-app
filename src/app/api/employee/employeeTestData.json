[
  {
    "id": 1,
    "name": "John",
    "email": "john@example.com",
    "status": "active"
  },
  {
    "id": 2,
    "name": "Ali",
    "email": "ali@example.com",
    "status": "active"
  },
  {
    "id": 3,
    "name": "Kelly",
    "email": "kelly@example.com",
    "status": "deactivated"
  }
]
