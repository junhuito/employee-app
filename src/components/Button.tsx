import { ButtonHTMLAttributes, DetailedHTMLProps } from 'react';

type ButtonDefaultHTMLAttributes = DetailedHTMLProps<
  ButtonHTMLAttributes<HTMLButtonElement>,
  HTMLButtonElement
>;

export type ButtonTheme = 'primary' | 'secondary' | 'danger';

type Button = {
  text: string;
  theme?: ButtonTheme;
  type?: 'button' | 'reset' | 'submit';
};

export const Button = (props: ButtonDefaultHTMLAttributes & Button) => {
  const {
    text,
    theme = 'primary',
    type = 'button',
    ...buttonDefaultHTMLAttributes
  } = props;

  const buttonTheme: Record<ButtonTheme, string> = {
    primary: 'bg-btn-primary-700 hover:bg-btn-primary-500',
    secondary: 'bg-btn-secondary-700 hover:bg-btn-secondary-500',
    danger: 'bg-btn-danger-700 hover:bg-btn-danger-500',
  };

  return (
    <button
      {...buttonDefaultHTMLAttributes}
      type={type}
      className={`text-white ${buttonTheme[theme]} cursor-pointer rounded-lg px-2 py-2.5 text-xs font-medium focus:outline-none disabled:cursor-not-allowed`}
    >
      {text}
    </button>
  );
};
