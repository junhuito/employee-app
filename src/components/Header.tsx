export const Header = () => {
  return (
    <nav className='border-gray-200 bg-primary-800 text-white'>
      <div className='mx-auto flex max-w-screen-xl flex-wrap items-center justify-between p-4'>
        <a
          href={'./'}
          className='flex items-center space-x-3 rtl:space-x-reverse'
        >
          <span className='bg-gradient-to-r from-sky-400 to-emerald-600 bg-clip-text text-2xl font-extrabold tracking-widest text-transparent'>
            AYP
          </span>
        </a>
      </div>
    </nav>
  );
};
