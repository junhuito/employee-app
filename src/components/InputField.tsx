import {
  DetailedHTMLProps,
  HTMLInputTypeAttribute,
  InputHTMLAttributes,
} from 'react';

type InputText = DetailedHTMLProps<
  InputHTMLAttributes<HTMLInputElement>,
  HTMLInputElement
>;

type InputField = InputText & {
  label: string;
  type?: HTMLInputTypeAttribute;
};

export const InputField = (props: InputField) => {
  const { label, type = 'text', ...inputProps } = props;

  return (
    <div>
      <label className='mb-2 block text-sm font-medium text-black dark:text-white'>
        {label}
      </label>
      <input
        {...inputProps}
        type={type}
        className='block w-full rounded-lg border-2 border-gray-200 p-2.5 text-sm text-black outline-none'
      />
    </div>
  );
};
