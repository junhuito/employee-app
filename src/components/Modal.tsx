import classnames from 'classnames';

type Modal = {
  children: JSX.Element;
  visible?: boolean;
  onDismiss?: () => void;
};

export const Modal = ({
  onDismiss = () => void 0,
  visible = false,
  children,
}: Modal) => {
  return (
    <div
      role='modal'
      onClick={onDismiss}
      className={`${classnames({
        hidden: !visible,
      })} fixed !inset-0 z-50 !m-0 flex items-center justify-center bg-black bg-opacity-60`}
    >
      <div onClick={(e) => e.stopPropagation()}>{children}</div>
    </div>
  );
};
