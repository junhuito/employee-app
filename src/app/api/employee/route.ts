import { NextRequest } from 'next/server';
import employeeTestData from './employeeTestData.json';

export async function GET(request: NextRequest) {
  return Response.json({ data: employeeTestData });
}
