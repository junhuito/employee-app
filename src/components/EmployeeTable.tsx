import { EmployeeStatus } from '@/constants/employee.enum';
import { Button } from './Button';
import { EmployeeData } from '@/interfaces/employee.interface';
import classNames from 'classnames';

export type EmployeeTable = {
  data: EmployeeData[];
  onActionClick?: (callback: EmployeeData) => void;
};

export const EmployeeTable = ({
  data = [],
  onActionClick = () => void 0,
}: EmployeeTable) => {
  return (
    <div className='relative overflow-x-auto rounded-lg border-2 shadow-md dark:border-none'>
      <table className='w-full text-left text-sm text-gray-500 rtl:text-right dark:text-gray-400 md:table-fixed'>
        <caption className='border-b-2 bg-white p-5 text-left text-lg font-semibold text-gray-900 rtl:text-right dark:border-none dark:bg-gray-800 dark:text-white'>
          Employees
        </caption>
        <thead className='uppercas border-b-2 bg-stone-50 text-xs text-gray-700 dark:border-none dark:bg-gray-700 dark:text-gray-400'>
          <tr>
            <th scope='col' className='px-6 py-3 capitalize'>
              ID
            </th>
            <th scope='col' className='px-6 py-3 capitalize'>
              Name & Email
            </th>
            <th scope='col' className='px-6 py-3 capitalize'>
              Status
            </th>
            <th scope='col' className='px-6 py-3 capitalize'>
              <span className='sr-only'>Action</span>
            </th>
          </tr>
        </thead>
        <tbody>
          {data.map((employeeData) => (
            <tr
              key={employeeData.id}
              className='border-b-2 bg-white text-xs hover:bg-gray-50 dark:border-gray-700 dark:bg-gray-800 dark:hover:bg-gray-600'
            >
              <td scope='row' className='whitespace-nowrap px-6 py-4 font-bold'>
                {employeeData.id}
              </td>
              <td className='px-6 py-4'>
                <p>{employeeData.name}</p>
                <p>{employeeData.email}</p>
              </td>
              <td className={`px-6 py-4 capitalize`}>
                <span
                  className={`rounded-md px-2 py-1 ${classNames({
                    ['border border-green-300 bg-green-100 text-green-800']:
                      employeeData.status === EmployeeStatus.ACTIVE,
                    ['border border-red-300 bg-red-100 text-red-800']:
                      employeeData.status !== EmployeeStatus.ACTIVE,
                  })}`}
                >
                  {employeeData.status}
                </span>
              </td>
              <td className='px-6 py-4 text-right'>
                <Button
                  text='Edit'
                  onClick={(e) => onActionClick(employeeData)}
                />
              </td>
            </tr>
          ))}
        </tbody>
      </table>
    </div>
  );
};
