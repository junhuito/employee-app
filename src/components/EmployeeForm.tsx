import { ChangeEvent, FormEvent, useState } from "react";
import { Button } from "./Button";
import { InputField } from "./InputField";
import { Toggle } from "./Toggle";

type EmployeeFormData = {
    name: string;
    email: string;
    status: boolean;
};

export const EmployeeForm = () => {

    const [employeeData, setEmployeeData] = useState<EmployeeFormData>({
        name: '',
        email: '',
        status: false,
    });

    const handleChange = (e: ChangeEvent<HTMLInputElement>) => {
        const { name, value } = e.target;
        setEmployeeData((prev) => ({ ...prev, [name]: value }))
    }

    const handleToggle = (status: boolean) => {
        setEmployeeData((prev) => ({ ...prev, status }))
    }

    const handleSubmit = (e: FormEvent<HTMLFormElement>) => {
        e.preventDefault();
    }

    return (
      <form
        onSubmit={handleSubmit}
        className="bg-white dark:bg-primary-700 p-10 rounded-lg space-y-5"
      >
        <InputField
          label="Name"
          name="name"
          onChange={handleChange}
          placeholder="name"
          required
          value={employeeData.name}
        />
        <InputField
          type="email"
          name="email"
          label="Email"
          onChange={handleChange}
          placeholder="johndoe@sample.com"
          required
        />
        <div className="block w-full">
          <Toggle label="Status" checked={employeeData.status} onChange={handleToggle}/>
        </div>
        <div className="flex justify-end">
          <Button type="submit" text="Submit" />
        </div>
      </form>
    );
}