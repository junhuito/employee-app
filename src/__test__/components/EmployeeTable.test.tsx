import { render } from '@testing-library/react';
import { EmployeeTable } from '@/components/EmployeeTable';
import { EmployeeStatus } from '@/constants/employee.enum';
import '@testing-library/jest-dom';

describe('EmployeeTable', () => {
  it('should render table with employee info', () => {
    const { container } = render(
      <EmployeeTable
        data={[
          {
            id: 1,
            name: 'test employee',
            email: 'test@example.com',
            status: EmployeeStatus.ACTIVE,
          },
          {
            id: 2,
            name: 'test employee 2',
            email: 'test2@example.com',
            status: EmployeeStatus.DEACTIVATED,
          },
        ]}
      />
    );

    expect(container).toMatchSnapshot();
  });
});
