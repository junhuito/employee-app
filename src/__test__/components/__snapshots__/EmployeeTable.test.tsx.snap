// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmployeeTable should render table with employee info 1`] = `
<div>
  <div
    class="relative overflow-x-auto rounded-lg border-2 shadow-md dark:border-none"
  >
    <table
      class="w-full text-left text-sm text-gray-500 rtl:text-right dark:text-gray-400 md:table-fixed"
    >
      <caption
        class="border-b-2 bg-white p-5 text-left text-lg font-semibold text-gray-900 rtl:text-right dark:border-none dark:bg-gray-800 dark:text-white"
      >
        Employees
      </caption>
      <thead
        class="uppercas border-b-2 bg-stone-50 text-xs text-gray-700 dark:border-none dark:bg-gray-700 dark:text-gray-400"
      >
        <tr>
          <th
            class="px-6 py-3 capitalize"
            scope="col"
          >
            ID
          </th>
          <th
            class="px-6 py-3 capitalize"
            scope="col"
          >
            Name & Email
          </th>
          <th
            class="px-6 py-3 capitalize"
            scope="col"
          >
            Status
          </th>
          <th
            class="px-6 py-3 capitalize"
            scope="col"
          >
            <span
              class="sr-only"
            >
              Action
            </span>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          class="border-b-2 bg-white text-xs hover:bg-gray-50 dark:border-gray-700 dark:bg-gray-800 dark:hover:bg-gray-600"
        >
          <td
            class="whitespace-nowrap px-6 py-4 font-bold"
            scope="row"
          >
            1
          </td>
          <td
            class="px-6 py-4"
          >
            <p>
              test employee
            </p>
            <p>
              test@example.com
            </p>
          </td>
          <td
            class="px-6 py-4 capitalize"
          >
            <span
              class="rounded-md px-2 py-1 border border-green-300 bg-green-100 text-green-800"
            >
              active
            </span>
          </td>
          <td
            class="px-6 py-4 text-right"
          >
            <button
              class="text-white bg-btn-primary-700 hover:bg-btn-primary-500 cursor-pointer rounded-lg px-2 py-2.5 text-xs font-medium focus:outline-none disabled:cursor-not-allowed"
              type="button"
            >
              Edit
            </button>
          </td>
        </tr>
        <tr
          class="border-b-2 bg-white text-xs hover:bg-gray-50 dark:border-gray-700 dark:bg-gray-800 dark:hover:bg-gray-600"
        >
          <td
            class="whitespace-nowrap px-6 py-4 font-bold"
            scope="row"
          >
            2
          </td>
          <td
            class="px-6 py-4"
          >
            <p>
              test employee 2
            </p>
            <p>
              test2@example.com
            </p>
          </td>
          <td
            class="px-6 py-4 capitalize"
          >
            <span
              class="rounded-md px-2 py-1 border border-red-300 bg-red-100 text-red-800"
            >
              deactivated
            </span>
          </td>
          <td
            class="px-6 py-4 text-right"
          >
            <button
              class="text-white bg-btn-primary-700 hover:bg-btn-primary-500 cursor-pointer rounded-lg px-2 py-2.5 text-xs font-medium focus:outline-none disabled:cursor-not-allowed"
              type="button"
            >
              Edit
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
